torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,193 @@
+        all_results = None
+        if self.is_coordinator:
+            node_failures = {i: err for i, err in enumerate(all_data) if isinstance(err, BaseException)}
+
+            if len(node_failures) == 0:
+                try:
+                    all_results = reduce_fun(all_data)
+                except BaseException as e:
+                    node_failures[self.rank] = e
+
+            if len(node_failures) > 0:
+                all_results = [CheckpointException(step, node_failures)] * self.get_world_size()
+
+        result = self.scatter_object(all_results)
+        if isinstance(result, CheckpointException):
+            raise result
+        return result
+
+    def all_reduce(self,
+        step: str,
+        map_cb: Callable[[], Any],
+        reduce_cb: Callable[[List[Any]], Any]
+    ) -> Tuple[Any, Any]:
+        try:
+            local_data = map_cb()
+        except BaseException as e:
+            local_data = e
+
+        all_data = self.gather_object(local_data)
+        result = None
+        if self.is_coordinator:
+            node_failures = {i: err for i, err in enumerate(all_data) if isinstance(err, BaseException)}
+
+            if len(node_failures) == 0:
+                try:
+                    result = reduce_cb(all_data)
+                except BaseException as e:
+                    node_failures[self.rank] = e
+
+            if len(node_failures) > 0:
+                result = CheckpointException(step, node_failures)
+
+        result = self.broadcast_object(result)
+        if isinstance(result, CheckpointException):
+            raise result
+        return result
+
+    def all_gather(self,
+        step: str,
+        map_fun: Callable[[], Any],
+    ) -> Tuple[Any, Any]:
+        try:
+            result = map_fun()
+        except BaseException as e:
+            result = e
+
+        all_results = self.all_gather_object(result)
+
+        node_failures = {i: err for i, err in enumerate(all_results) if isinstance(err, BaseException)}
+        if len(node_failures) > 0:
+            raise CheckpointException(step, node_failures)
+        return all_results
+
+    def broadcast(self,
+        step: str,
+        map_cb: Callable[[], Any],
+    ) -> Any:
+        result = None
+        if self.is_coordinator:
+            try:
+                result = map_cb()
+            except BaseException as e:
+                result = CheckpointException(step, { self.rank: e })
+        result = self.broadcast_object(result)
+        if isinstance(result, CheckpointException):
+            raise result
+        return result
+
+    def all_to_all(self,
+        step: str,
+        map_cb: Callable[[], Any],
+    ) -> Any:
+        result = None
+        if self.is_coordinator:
+            try:
+                result = map_cb()
+            except BaseException as e:
+                result = CheckpointException(step, { self.rank: e })
+        result = self.broadcast_object(result)
```

Comment:
this looks wrong? it seems needing all_to_all

Reply:
this was deadcode, removed it.